lib/interface_quda.cpp

```diff
@@ -2205,70 +2205,123 @@ void cloverQuda(void *h_out, void *h_in, QudaInvertParam *inv_param, QudaParity
+  // for M, Mdag, MdagM, and MMdag type operators, chosen via
+  // eig_use_dagger and eig_use_norm_op booleans,
+  // each combination of which may be preconditioned via eig_use_pc_op. We select
+  // the correct QudaInvertParam values for the solve_type and
+  // solution_type based on those three booleans
+
+  if (eig_param->use_pc) {
+    if (eig_param->use_norm_op)
+      inv_param->solve_type = QUDA_NORMOP_PC_SOLVE;
+    else
+      inv_param->solve_type = QUDA_DIRECT_PC_SOLVE;
+    inv_param->solution_type = QUDA_MATPC_SOLUTION;
+  } else {
+    if (eig_param->use_norm_op)
+      inv_param->solve_type = QUDA_NORMOP_SOLVE;
+    else
+      inv_param->solve_type = QUDA_DIRECT_SOLVE;
+    inv_param->solution_type = QUDA_MAT_SOLUTION;
   }
+  //------------------------------------------------------------------
 
+  // Ensure that the parameter structures are sound.
   checkInvertParam(inv_param);
   checkEigParam(eig_param);
-  cudaGaugeField *cudaGauge = checkGauge(inv_param);
 
-  bool pc_solve = (inv_param->solve_type == QUDA_DIRECT_PC_SOLVE) || (inv_param->solve_type == QUDA_NORMOP_PC_SOLVE)
-    || (inv_param->solve_type == QUDA_NORMERR_PC_SOLVE);
+  // Check that the gaige field is valid
+  cudaGaugeField *cudaGauge = checkGauge(inv_param);
 
+  // Set all timing statistics to zero
   inv_param->secs = 0;
   inv_param->gflops = 0;
   inv_param->iter = 0;
 
+  // Dump all eigensolver and invert param variables to stdout if requested.
+  pushVerbosity(inv_param->verbosity);
+  if (getVerbosity() >= QUDA_DEBUG_VERBOSE) {
+    printQudaInvertParam(inv_param);
+    printQudaEigParam(eig_param);
+  }
+
   // Define problem matrix
   //------------------------------------------------------
   Dirac *d = nullptr;
   Dirac *dSloppy = nullptr;
   Dirac *dPre = nullptr;
 
   // Create the dirac operator with a sloppy and a precon.
+  bool pc_solve = (inv_param->solve_type == QUDA_DIRECT_PC_SOLVE) || (inv_param->solve_type == QUDA_NORMOP_PC_SOLVE);
   createDirac(d, dSloppy, dPre, *inv_param, pc_solve);
   Dirac &dirac = *d;
+  //------------------------------------------------------
 
-  // Create device side ColorSpinorField vector space and to pass to the
-  // compute function.
-  const auto X = cudaGauge->X();
-  ColorSpinorParam cpuParam(host_evecs[0], *inv_param, X, inv_param->solution_type, inv_param->input_location);
-
-  // create wrappers around application vector set
-  std::vector<ColorSpinorField *> host_evecs_(eig_param->n_conv);
-  for (int i = 0; i < eig_param->n_conv; i++) {
+  // Construct vectors
+  //------------------------------------------------------
+  // Create host wrappers around application vector set
+  ColorSpinorParam cpuParam(host_evecs[0], *inv_param, cudaGauge->X(), inv_param->solution_type,
+                            inv_param->input_location);
+
+  int n_eig = eig_param->n_conv;
+  if (eig_param->compute_svd) n_eig *= 2;
+  std::vector<ColorSpinorField *> host_evecs_(n_eig);
```

Comment:
While reading this, I was thinking to suggest modifying this to vector of fields, rather than vector of objects.  But on second though, I think it makes more sense to get this PR in with all the tests and only then to perform surgery of that nature.